### PR TITLE
Add continuous integration on Travis

### LINKED ISSUE #40: Continuous Integration
We should set something on travis-CI.  Since we need a somewhat custom cross-compiler and qemu, we might want to use a docker file to set up everything we need

### PR DESCRIPTION
**Pull request by cstanfill**
@allanlw should check this Dockerfile before this gets merged.

Fixes #40

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
+sudo: true
```
Comment (allanlw) [inline .travis.yml]: This is a good script

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,10 @@
+FROM debian:stable
+WORKDIR /root/packages
+RUN apt-get update -qq && \
+    apt-get install -y build-essential wget git && \
+    apt-get install -y qemu-system-x86 && \
+    wget https://s3-us-west-2.amazonaws.com/colin.cool/gcc-cross-deb/binutils-x86-64-elf_2.28-5%2B1_amd64.deb && \
+    wget https://s3-us-west-2.amazonaws.com/colin.cool/gcc-cross-deb/gcc-x86-64-elf_6.3.0-18%2B1_amd64.deb && \
```
Comment (allanlw) [inline Dockerfile]: Anyway we could remove the dependency here?

File: Makefile [hunk 3]
```diff
@@ -117,8 +117,10 @@ userland/%/test_disk:
 ## But doesn't do anything else
 userland/%/coverage.log: userland/%/output.txt
 
+# Runs without a GDB stub so make -j works properly.
 userland/%/output.txt: userland/%.bin george.multiboot bootloader.multiboot userland/%/test_disk
 	./script/QEMU.sh \
+		--no-gdb \
```
Comment (allanlw) [inline Makefile]: oh lol oops this is good

File: script/QEMU.sh [hunk 4]
```diff
@@ -66,9 +73,15 @@ QEMU_ARGS+=" -drive file=$DRIVE,index=0,media=disk,format=raw"
...
 QEMU_ARGS+=" -device isa-debug-exit,iobase=0xf4,iosize=0x04"
 # Expose standard gdb on :1234
-QEMU_ARGS+=" -s"
+if [ "$GDB" -gt 0 ]; then
+  QEMU_ARGS+=" -s"
+fi
 # Pause on startup (require 'c' on console, or 'c' in GDB to start)
 if [ "$PAUSE" -gt 0 ]; then
+  if [ "$GDB" -eq 0 ]; then
+    echo "Error: -S requires a gdb stub to do anything."
```
Comment (allanlw) [inline script/QEMU.sh]: technically this isn't true, but it's not actually *useful* if you don't have the gdb stub.

you can press 'c' in the qemu console to continue execution.

Comment (gsilvis) [inline script/QEMU.sh]: I could theoretically imagine wanting to start up qemu but have it wait for me to hit 'c' before actually booting.  It takes /two/ manual flags to trigger this case, so I can't imagine doing it accidentally.  I guess I'd vaguely vote for allowing it.

File: script/QEMU.sh [hunk 5]
```diff
@@ -28,7 +29,10 @@ while true; do
       echo "  -S, --wait          Pause CPU on startup"
       echo "  -s, --serial=FILE   Output serial to FILE instead of stdout"
       echo "  -c, --coverage=FILE Output unique hit kernel mode %rip values to FILE"
+      echo "  -k, --enable-kvm    Enable hardware acceleration via KVM"
       echo "      --nographic     Don't display the graphics (default is SDL or curses)"
+      echo "      --no-reboot     Don't reboot the machine on hardware faults"
+      echo "      --no-gdb        Don't expose a gdb stub (useful if running in parallel)"
```
Comment (allanlw) [inline script/QEMU.sh]: This slipped by me the other time, but these `--no` flags don't follow the same convention.

File: Dockerfile (hunk 2 above)
Comment (cstanfill) [inline Dockerfile]: which one? binutils and gcc? something somewhere needs to actually build the kernel....

File: Makefile (hunk 3 above)
Comment (cstanfill) [inline Makefile]: ;)

File: script/QEMU.sh (hunk 4 above)
Comment (cstanfill) [inline script/QEMU.sh]: That's cool by me

File: script/QEMU.sh (hunk 5 above)
Comment (cstanfill) [inline script/QEMU.sh]: yeah, blame QEMU

**Comment by cstanfill** [comment]
PTAL I removed my flag check and renamed the nographic flag

File: Dockerfile (hunk 2 above)
Comment (allanlw) [inline Dockerfile]: I meant the dependency on an S3 bucket you own.

Comment (cstanfill) [inline Dockerfile]: I'm open to ideas. FWIW the packages are 2.1MB and 21MB respectively.

**Pull request merged**